keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,401 @@
+
+Beyond the above, there are no Kubernetes version skew considerations in the context of this KEP.
+
+## Implementation History
+
+07/16/2019: Initial KEP drafted
+
+## Drawbacks
+
+The main drawback associated with the approach leveraging csi-proxy.exe is that the life cycle of that binary will need to be managed out-of-band from Kubernetes. However, cluster administrators need to maintain and manage life cycle of other core binaries like kubeproxy.exe, kubelet.exe, dockerd.exe and containerd.exe (in the future). Therefore csi-proxy.exe will be one additional binary that will need to be treated in a similar way.
+
+The API versioning scheme described above, as of now, will maintain backward compatibility indefinitely. This requires the scope of csi-proxy.exe to be limited to a very scoped down fundamental set of operations. Maintainers therefore will need to be very cautious when accepting suggestions for new APIs and enhancements. This may slow progress at times.
+
+There may ultimately be certain operations that csi-proxy.exe cannot support in an elegant fashion and requires the plugin author targeting Windows nodes to seek one of the alternatives described below. This KEP and maintainers should ensure such instances are as limited as possible.
+
+## Alternatives
+
+There are multiple alternatives to deploying CSI Node Plugins as containers along with csi-proxy.exe for privileged operations however each has it's own set of disadvantages.
+
+### Deploying CSI Node Plugins as binaries and deployed as processes running on the host:
+
+With this approach, lifecycle of multiple stand-alone binaries corresponding to different CSI node plugins will need to be managed. The standard CSI Node Driver Registrar which is distributed as a container will also need to be repackaged as binaries and distributed (as mixing side car containers with standalone binaries is not possible). Managing several binaries outside of Kubernetes may not scale well as diverse storage systems, each with their own CSI plugin, is integrated with the cluster.
+
+### Package CSI Node Plugins as containers and deployed as processes running on the host:
+
+With this approach, the container run time is enhanced to be able to launch binaries directly on the host after pulling down a container image and extracting the binary from the image. While usual Kubernetes constructs may be used to launch pods with a special RuntimeClass that can handle launching of the binaries as processes on hosts, Kubernetes won't be able to fully monitor and control the whole life-cycle of the binaries post launch. Collection of logs from the plugins also become problematic and will require tooling out-of-band from Kubernetes.
```

Comment:
I think managing lifecycles and logs is possible, but more require work in the container runtime.

The cons mentioned here (i.e., lifecycle management and logs) also present in the approach above "Deploying CSI Node Plugins as binaries and deployed as processes running on the host".